commons-numbers-complex/src/main/java/org/apache/commons/numbers/complex/ComplexConstructor.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.commons.numbers.complex;
+
+/**
+ * Define a constructor for a Complex.
+ * @param <R> Generic
+ */
+@FunctionalInterface
+public interface ComplexConstructor<R> {
```

Comment:
The word "constructor" has a very specific meaning in Java. Perhaps we could use "ComplexFactory" here instead?

Reply:
I agree the term constructor is not ideal. The function is acting as both a Consumer (to receive the complex result) and optionally a Function (to convert (a +ib) to z). In both cases the operation is a terminating operation where the primitive (r, i) pair finally gets used and potentially converted to an object. So there is a lot to cover in a name. Switching to Factory implies that the function will always make something (as does constructor). However some use cases could return `Void`. So I think this also needs to be captured in the name. Perhaps ComplexResult?

Reply:
What do you think of using the name pair "source"/"sink". I could then picture the following:
```java
// interface for obtaining complex values
public interface ComplexSource {
    double getReal();
    double getImaginary();

    default real() { return getReal(); }
    default imag() { return getImaginary(); }
}

// interface for accepting complex values, optionally returning a result
@FunctionalInterface
public interface ComplexSink<R> {
    R accept(double real, double imaginary);
}
```
These interfaces could then be used as follows:
```java
// existing Complex class
public final class Complex implements ComplexSource {
    // ...
}

// TBD complex value container class
public class ComplexBuffer implements ComplexSource, ComplexSink<ComplexBuffer> {
    // indexing methods similar to java.nio.Buffer
    public int position();
    public ComplexBuffer position(int pos);

    // get the values at the current position
    public double getReal() { ... }
    public double getImaginary() { ... }

    // set the values at the current position
    ComplexBuffer accept(double real, double imaginary) {
        // ...
        return this;
    }
}
```